Fix log records for Kafka responses

tgraf opened the pull request.

rlenglet commented on pkg/kafka/correlation_cache.go:
```diff
@@ -0,0 +1,129 @@
+)
+
+const (
+	requestLifetime = time.Duration(2) * time.Minute
```
```
requestLifetime = 2 * time.Minute
```

tgraf replied:
fixed

rlenglet commented on pkg/kafka/correlation_cache.go:
```diff
@@ -0,0 +1,129 @@
+		}
+		cc.mutex.Unlock()
+
+		time.Sleep(1 * time.Minute)
```
You could determine the next deadline in the loop above, and wait until then.

tgraf replied:
That would certainly optimise on the most accurate expiration of entries but it causes a lot of churn as we pretty much run the GC for each individual entry that is about to expire.

rlenglet commented on pkg/kafka/correlation_cache.go:
```diff
@@ -0,0 +1,129 @@
+
+		cc.mutex.Lock()
+		for correlationID, req := range cc.cache {
+			if req.created.After(expiryWatermark) {
```
This will never be true. `req.created` is always <= `time.Now()`, so a fortiori it is always < `time.Now().Add(requestLifetime)`.
Shouldn't the logic be:
```
expiryWatermark := time.Now().Add(-requestLifetime)
...
if req.created.Before(expiryWatermark) {
```

Would it be possible to add a unit test for the GC logic? You might have to make the `requestLifetime` configurable.

tgraf replied:
fixed

rlenglet commented on pkg/proxy/accesslog/builder.go:
```diff
@@ -0,0 +1,265 @@
+	"strconv"
+	"time"
+
+	"github.com/cilium/cilium/pkg/identity"
```
`pkg/proxy/accesslog` is a public API, and is used by third-party software. Adding this file adds packages dependencies to those third-parties. Currently, the dependencies are intentionally relatively minimal:
```
$ grep -r github.com/cilium/cilium pkg/proxy/accesslog
pkg/proxy/accesslog/log.go:	"github.com/cilium/cilium/pkg/flowdebug"
pkg/proxy/accesslog/log.go:	"github.com/cilium/cilium/pkg/lock"
pkg/proxy/accesslog/log.go:	"github.com/cilium/cilium/pkg/logging"
pkg/proxy/accesslog/log.go:	"github.com/cilium/cilium/pkg/logging/logfields"
```

Could it be possible to move this file out of the `pkg/proxy/accesslog` package?

tgraf replied:
Agreed, I'll move it into a subpackage

tgraf replied:
Fixed. I've moved all of `accesslog/log.go` out so the dependencies are less than before.

rlenglet commented on pkg/proxy/accesslog/log.go:
```diff
@@ -22,6 +22,7 @@ import (
 	"github.com/cilium/cilium/pkg/logging"
 	"github.com/cilium/cilium/pkg/logging/logfields"
 
+	"github.com/sirupsen/logrus"
```
Same comment as for `builder.go`, this is imposing a new dependency on third-parties using the `pkg/proxy/accesslog` package. Could this be avoided?

rlenglet commented on pkg/kafka/correlation_cache.go:
```diff
@@ -0,0 +1,175 @@
+// function must be called when the cache is no longer required, otherwise go
+// routines are leaked.
+func (cc *CorrelationCache) DeleteCache() {
+	close(cc.stopGc)
```
Nit. This can be made idempotent:
```
	select {
	case <-cc.stopGc:
	default:
		close(cc.stopGc)
	}
```

rlenglet commented on pkg/kafka/correlation_cache.go:
```diff
@@ -0,0 +1,175 @@
+
+		cc.mutex.Lock()
+		for correlationID, entry := range cc.cache {
+			if entry.created.After(expiryCreationTime) {
```
After -> Before

tgraf replied:
Fixed

rlenglet commented on pkg/kafka/correlation_cache_test.go:
```diff
@@ -0,0 +1,74 @@
+	cc := NewCorrelationCache()
+	cc.Insert(request1, nil)
+
+	// wait for the garbage collector to expire the request
```
You need to test also that more recent entries are not GCed. I'd try this test scenario:

1. Set the lifetime to 200ms. Create the cache + start the GC.
2. Sleep 100ms, to wait for the initial GC run to complete.
3. Insert request1.
4. Sleep 200ms, to let the GC run.
5. Check that **request1 has not been GCed**.
6. Insert request2.
7. Sleep 200ms, to let the GC run again.
8. Check that **request1 has been GCed**, and that **request2 has not been GCed**.

This scenario should fail with your current implementation, and pass after fixing After -> Before above.

tgraf replied:
Fixed, I've introduced a gc run counter to make this reliable

manalibhutiyani commented on pkg/proxy/kafka.go:
```diff
@@ -246,6 +251,9 @@ func (k *kafkaRedirect) handleRequest(pair *connectionPair, req *kafka.RequestMe
 		return
 	}
 
+	// The request is allowd so we add the request to the correlation cache
```
Nit: allowed

tgraf replied:
Fixed

rlenglet commented on pkg/proxy/proxy.go:
```diff
@@ -111,7 +111,35 @@ func newRedirect(port uint16, source ProxySource, id string) *Redirect {
 }
 
-func (r *Redirect) updateAccounting(t accesslog.FlowType, verdict accesslog.FlowVerdict) {
+func (r *Redirect) DeriveEndpointInfo(ip net.IP, info *accesslog.EndpointInfo) bool {
```
Remove. This doesn't seem to be called anywhere.

tgraf replied:
Did I miss to commit something?

```
pkg/proxy/logger/logger.go-             // if IP belongs to the cluster.
pkg/proxy/logger/logger.go-             if node.GetIPv4ClusterRange().Contains(ip) {
pkg/proxy/logger/logger.go-                     // If endpoint cannot be found, set to cluster identity
pkg/proxy/logger/logger.go:                     if !producer.DeriveEndpointInfo(ip, info) {
pkg/proxy/logger/logger.go-                             fillIdentity(info, identity.ReservedIdentityCluster)
pkg/proxy/logger/logger.go-                     }
pkg/proxy/logger/logger.go-             } else {
--
pkg/proxy/logger/logger.go-             }
pkg/proxy/logger/logger.go-
pkg/proxy/logger/logger.go-             if node.GetIPv6ClusterRange().Contains(ip) {
pkg/proxy/logger/logger.go:                     if !producer.DeriveEndpointInfo(ip, info) {
pkg/proxy/logger/logger.go-                             fillIdentity(info, identity.ReservedIdentityCluster)
pkg/proxy/logger/logger.go-                     }
pkg/proxy/logger/logger.go-             } else {
```

rlenglet commented on pkg/proxy/proxy.go:
```diff
@@ -111,7 +111,35 @@ func newRedirect(port uint16, source ProxySource, id string) *Redirect {
 
-func (r *Redirect) updateAccounting(t accesslog.FlowType, verdict accesslog.FlowVerdict) {
+func (r *Redirect) DeriveEndpointInfo(ip net.IP, info *accesslog.EndpointInfo) bool {
+	if ep := endpointmanager.LookupIPv4(addressing.DeriveCiliumIPv4(ip).String()); ep != nil {
```
Does this assume that IPv4 is always enabled?

tgraf replied:
Yes, this is existing code. We should fix it up.

rlenglet commented on pkg/proxy/proxy.go:
```diff
@@ -111,7 +111,35 @@ func newRedirect(port uint16, source ProxySource, id string) *Redirect {
+
+// GetObservationPoint returns the observation point at which the redirect is
+// attached to
+func (r *Redirect) GetObservationPoint() accesslog.ObservationPoint {
```
Remove. This doesn't seem to be called anywhere.

tgraf replied:
```
pkg/proxy/logger/logger.go-     lr := LogRecord{
pkg/proxy/logger/logger.go-             LogRecord: accesslog.LogRecord{
pkg/proxy/logger/logger.go-                     Type:              t,
pkg/proxy/logger/logger.go:                     ObservationPoint:  producer.GetObservationPoint(),
pkg/proxy/logger/logger.go-                     IPVersion:         accesslog.VersionIPv4,
pkg/proxy/logger/logger.go-                     TransportProtocol: 6,
pkg/proxy/logger/logger.go-                     Timestamp:         time.Now().UTC().Format(time.RFC3339Nano),
```

rlenglet commented on pkg/proxy/proxy.go:
```diff
@@ -111,7 +111,35 @@ func newRedirect(port uint16, source ProxySource, id string) *Redirect {
+
+// UpdateAccounting is called for each log record emitted as soon as the
+// verdict is known
+func (r *Redirect) UpdateAccounting(t accesslog.FlowType, verdict accesslog.FlowVerdict) {
```
Remove. This doesn't seem to be called anywhere anymore.

tgraf replied:
```
pkg/proxy/logger/logger.go-             lr.Verdict = v
pkg/proxy/logger/logger.go-             lr.Info = info
pkg/proxy/logger/logger.go-
pkg/proxy/logger/logger.go:             lr.producer.UpdateAccounting(lr.Type, lr.Verdict)
pkg/proxy/logger/logger.go-     }
pkg/proxy/logger/logger.go-}
pkg/proxy/logger/logger.go-
```

rlenglet commented on pkg/proxy/proxy.go:
```diff
@@ -651,3 +507,20 @@ func (p *Proxy) GetStatusModel() *models.ProxyStatus {
 }
+
+// LocalEndpointInfo return an EndpointInfo with the information of the local endpoint
+func (r *Redirect) LocalEndpointInfo() accesslog.EndpointInfo {
```
Remove. This doesn't seem to be called anywhere anymore.

tgraf replied:
```
pkg/proxy/logger/logger.go-     return func(lr *LogRecord) {
pkg/proxy/logger/logger.go-             switch lr.ObservationPoint {
pkg/proxy/logger/logger.go-             case accesslog.Ingress:
pkg/proxy/logger/logger.go:                     lr.DestinationEndpoint = lr.producer.LocalEndpointInfo()
pkg/proxy/logger/logger.go-             case accesslog.Egress:
pkg/proxy/logger/logger.go:                     lr.SourceEndpoint = lr.producer.LocalEndpointInfo()
pkg/proxy/logger/logger.go-             }
pkg/proxy/logger/logger.go-
pkg/proxy/logger/logger.go-             ipstr, port, err := net.SplitHostPort(i.SrcIPPort)
```

manalibhutiyani commented on pkg/proxy/kafka.go:
```diff
@@ -332,20 +346,21 @@ func (k *kafkaRedirect) handleResponses(done <-chan struct{}, pair *connectionPa
 		}
 
+		// Find the request that correlates with this response based on the
+		// correlation ID
```
I am not sure if `correlation ID` is enough to say that a particular response belongs to a request. `correlation ID` is a user specified field. So 2 requests with different topics can have the same correlation ID back to back. We might not do the correct thing in that scenario. Would it make sense to have both correlation ID as well as topic /client ID to correlate request as well as response?

manalibhutiyani replied:
Also, correlation ID is not a mandatory field, so it may/may not be zero. Not sure how this will work in that case. Please correct me if I am mistaken?

tgraf replied:
Good point. We can also rely on order + correlation id. What worries me is the following in the spec:

> All requests are initiated by the client, and result in a corresponding response message from the server except where noted.

I haven't found any details in the spec on when exactly that is the case.

tgraf replied:
Discussed more with @raybejjani. He brought up the idea that we could modify the correlationID and set it back in the response direction. This seems the most reliable way to go forward so I will implement this.

manalibhutiyani reviewed (commented):
Review for commit: kafka: Correlate response with request to generate log record

manalibhutiyani commented on pkg/kafka/correlation_cache.go:
```diff
@@ -0,0 +1,175 @@
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+var (
```
Would be great to have a comment here describing `requestLifetime` in relation with `correlation cache`

tgraf committed:
kafka: Add GetCorrelationID() accessor for responses

This will be required to implement a correlation cache

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
accesslog: Document EndpointInfo

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
kafka : Cleanup use of identity package

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
envoy: Provide accessor functions for accesslog needs

This provides compatibility to pkg/proxy/accesslog types and is in preparation
for an overall proxy log cleanup

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
proxy: Move all generic logging code into pkg/proxy/logger

A lot of code was not shared between the Kafka proxy and the Envoy
management layer. This makes it more difficult to support additional
proxies in the future.

Introduce a set of new functions:

 - logger.NewLogRecord(producer, type, ...tags) LogRecord
 - LogRecord.ApplyTags(...tags)
 - LogRecord.Log()

Example:
``` go
	record := accesslog.NewLogRecord(r.redirect, pblog.GetFlowType(),
		accesslog.LogTags.Timestamp(time.Unix(int64(pblog.Timestamp/1000000000), int64(pblog.Timestamp%1000000000))),
		accesslog.LogTags.Verdict(pblog.GetVerdict(), pblog.CiliumRuleRef),
		accesslog.LogTags.Addressing(accesslog.AddressingInfo{
			SrcIPPort:   pblog.SourceAddress,
			DstIPPort:   pblog.DestinationAddress,
			SrcIdentity: pblog.SourceSecurityId,
		}),
		accesslog.LogTags.HTTP(&accesslog.LogRecordHTTP{
			Method: pblog.Method,
			Code:   int(pblog.Status),
[...]
```

The producer has to provide the following functionality:
``` go
type LogRecordProducer interface {
	// DeriveEndpointInfo must lookup the endpoint by IP and fill the
	// endpoint's parameters into the EndpointInfo struct. The function
	// must return true if the endpoint was found.
	DeriveEndpointInfo(ip net.IP, info *EndpointInfo) bool

	// LocalEndpoint must return information about the local endpoint which
	// either sent the request (for egress) or is receiving the request
	// (for ingress)
	LocalEndpointInfo() EndpointInfo

	// GetObservationPoint must return the ObservationPoint at which the log record is being produced
	GetObservationPoint() ObservationPoint

	// UpdateAccounting is called for each log record as soon as the log
	// verdict is known
	UpdateAccounting(t FlowType, v FlowVerdict)
}
```

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
kafka: Correlate response with request to generate log record

The existing code passed the request log record of the initial request of the
connection into the go routine to handle responses. This initial request was
then used to fill in fields such as topic when responses were logged. This is
incorrect as a single connection will transport many requests on different
topics with responses being received possibly in random order.

Introduce a correlation cache for Kafka request and response messages.

This will also allow to measure the time span between the request and response
for tracing use cases.

Signed-off-by: Thomas Graf <thomas@cilium.io>

manalibhutiyani commented on pkg/kafka/correlation_cache.go:
```diff
@@ -0,0 +1,224 @@
+//   request.
+//
+// A garbage collector will run frequently and expire requests which have not
+// been correlate for the period of `RequestLifetime`
```
nit: `correlated`?

tgraf replied:
fixed

manalibhutiyani commented on pkg/kafka/correlation_cache.go:
```diff
@@ -0,0 +1,224 @@
+// correlates the corresponding request, restores the original correlation ID
+// in the response and returns the original request
+func (cc *CorrelationCache) CorrelateResponse(res *ResponseMessage) *RequestMessage {
+	entry := cc.correlate(res.GetCorrelationID())
```
I am wondering if there can be a possible race here:
i.e between `correlate` which correlates an entry and `delete` which is suppose to delete that entry?

Scenario:
Request 1: Correlation ID : 5 
correlate return 5 for request 1
Request 2: Correlation ID : 6
cc.cache overflow, so correlation ID 5 gets overwritten with 6
cc.Delete now goes and deletes 6 instead of 5, since it only checks for correlation ID and not request.


Should be make the looking up and deleting a single function protected by a lock?

Let me know if this scenario cannot happen.

tgraf replied:
I don't follow regarding the race. `Delete()` was called on the correlation ID of the modified request. I've changed the code to get rid of `Delete()`, this might make it easier to understand.

manalibhutiyani commented on pkg/kafka/correlation_cache.go:
```diff
@@ -0,0 +1,224 @@
+		select {
+		case <-cc.stopGc:
+			return
+		default:
```
This might be a `go specific thing` which I don't understand, but why is default needed here?

tgraf replied:
Writing it this way makes it the most obvious to me that we are checking if the channel has been closed without blocking.

manalibhutiyani commented on pkg/proxy/accesslog/record.go:
```diff
@@ -68,15 +68,34 @@ const (
 
-// EndpointInfo contains information about the endpoint sending/receiving the flow
+// EndpointInfo is used for the SourceEndpoint and DestinationEndpoint field
+// and contains information about the sending respectively receiving endpoint.
```
Nit: The wording is a bit unclear to me
`sending respectively receiving endpoint.`

manalibhutiyani commented on pkg/proxy/kafka.go:
```diff
@@ -106,115 +106,113 @@ func createKafkaRedirect(r *Redirect, conf kafkaConfiguration) (RedirectImplemen
+			logger.LogTags.Kafka(&accesslog.LogRecordKafka{
+				APIVersion:    req.GetVersion(),
+				APIKey:        apiKeyToString(req.GetAPIKey()),
+				CorrelationID: int32(req.GetCorrelationID()),
```
Should we also be filling `ClientID` if present in the request?

tgraf replied:
I'm not opposed to adding the ClientID to the log record but it's outside of the scope of this PR.

manalibhutiyani commented on pkg/proxy/kafka.go:
```diff
@@ -106,115 +106,113 @@ func createKafkaRedirect(r *Redirect, conf kafkaConfiguration) (RedirectImplemen
+	if req != nil {
+		lr.Kafka.APIVersion = req.GetVersion()
+		lr.Kafka.APIKey = apiKeyToString(req.GetAPIKey())
+		lr.topics = req.GetTopics()
```
Ditto, should we be filling clientID in response if present in corresponding request

manalibhutiyani reviewed (commented):
Overall kudos to the effort. The code looks amazing!

This is my first pass of the review comments. I have focussed on the Kafka bits. Will try and go over other bits of code in the evening.

manalibhutiyani reviewed (approved):
Looks great!
